rust/otap-dataflow/crates/pdata/src/views/otlp/bytes/decode.rs

```diff
@@ -326,27 +326,48 @@ where
     }
 }
 
+/// This is a helper trait for adapting iterators of primitive fields using the packed encoding
+/// so that they can be used generically in [`RepeatedPrimitiveIter`]
+trait PackedIter<'a> {
+    type DecodedValue;
+
+    fn new(buffer: &'a [u8]) -> Self;
+
+    fn decode_value(slice: &[u8]) -> Option<Self::DecodedValue>;
+}
+
 /// Iterator for producing elements whose type is a repeated primitive where that primitive
-/// can be encoded as a fixed64. e.g. `repeated double` or `repeated fixed64`. OTLP uses the
-/// packed encoding for these types of fields
-/// https://protobuf.dev/editions/features/#repeated_field_encoding
+/// can be encoded as a fixed64. e.g. `repeated double` or `repeated fixed64` and the field
+/// is encoded using packed encoding.
+///
+/// Note this is for internal use only. The proto documentation states that decoders should
+/// have the flexibility to accept both packed and expanded encodings. Unless we're sure that
+/// the field is packed encoded, it's safer to use [`RepeatedFixed64Iter`] which automatically
+/// handles both encodings.
 pub struct PackedFixed64Iter<'a, V: FromFixed64> {
```

Comment:
Not for this PR, but we could consider implementing [`ExactSizeIterator`](https://doc.rust-lang.org/std/iter/trait.ExactSizeIterator.html) trait for this struct to make the `collect` calls from the consumer of this struct more efficient.